src/Google.Cloud.Functions.Invoker/HostingInternals.cs

```diff
@@ -0,0 +1,203 @@
+
+namespace Google.Cloud.Functions.Invoker
+{
+    internal static class HostingInternals
+    {
+        private static readonly IDictionary<string, string> CommandLineMappings = new Dictionary<string, string>
+        {
+            { "--target", $"{FunctionsFrameworkOptions.ConfigSection}:{nameof(FunctionsFrameworkOptions.FunctionTarget)}" },
+            { "--port", $"{FunctionsFrameworkOptions.ConfigSection}:{nameof(FunctionsFrameworkOptions.Port)}" },
+        };
+
+        internal static IConfigurationBuilder AddCommandLineArguments(IConfigurationBuilder configBuilder, string[] args)
+        {
+            if (args is null || args.Length == 0)
+            {
+                return configBuilder;
+            }
+            if (args.Length == 1)
+            {
+                if (args[0] == "%LAUNCHER_ARGS%")
+                {
+                    throw new InvalidOperationException("Unable to launch Web SDK project with launch settings. Please see https://github.com/GoogleCloudPlatform/functions-framework-dotnet/blob/master/docs/launch-settings.md");
+                }
+                args = new[] { "--target", args[0] };
+            }
+            configBuilder.AddCommandLine(args, CommandLineMappings);
+            return configBuilder;
+        }
+
+        internal static void ConfigureApplication(WebHostBuilderContext context, IApplicationBuilder app)
+        {
+            var configurers = app.ApplicationServices.GetServices<FunctionsStartup>();
+            foreach (var configurer in configurers)
+            {
+                configurer.Configure(context, app);
+            }
+
+            app.Map("/robots.txt", ReturnNotFound);
+            app.Map("/favicon.ico", ReturnNotFound);
+            app.Run(Execute);
+
+            // Slight hack using dependency injection to propagate the original function from ConfigureFunctionsFrameworkTarget
+            // to here.
+            var functionType = app.ApplicationServices.GetRequiredService<FunctionTypeProvider>().FunctionType;
+            // Note: we can't use ILogger<EntryPoint> as EntryPoint is static. This is an equivalent.
+            app.ApplicationServices
+                .GetRequiredService<ILoggerFactory>()
+                .CreateLogger(typeof(EntryPoint).FullName)
+                .LogInformation($"Serving function {functionType.FullName}");
+        }
+
+        internal static IEnumerable<FunctionsStartup> GetStartups(Assembly assembly) =>
+            assembly
+                .GetCustomAttributes<FunctionsStartupAttribute>()
+                .OrderBy(attr => attr.Order)
+                .ThenBy(attr => attr.StartupType.FullName)
+                .Select(attr => Activator.CreateInstance(attr.StartupType))
+                .Cast<FunctionsStartup>();
+
+        internal static void AddStartup(WebHostBuilderContext context, IServiceCollection services, FunctionsStartup startup)
+        {
+            startup.ConfigureServices(context, services);
```

Comment:
I think this should be done in ConfigureApplication. The user might be trying to override some services and when doing it here, it's us who might end overriding theirs. We "promised" to execute the Startups at the end, on the UseFunctionsFramework extension, but we are setting the services right here instead. Else add a note on the public extensions that add startups that the services will be added inmediately but the midlewares will be added later.